ledger/streaming/client/src/backfill.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A BlockStream that backfills missing indices from a given BlockFetcher.
+
+use displaydoc::Display;
+use futures::{FutureExt, Stream, StreamExt};
+use mc_common::logger::{log, Logger};
+use mc_ledger_streaming_api::{BlockFetcher, BlockStream, BlockStreamComponents, Result};
+use mc_transaction_core::BlockIndex;
+use std::{pin::Pin, sync::Arc};
+
+/// A [BlockStream] that backfills another [BlockStream] using a [BlockFetcher].
+#[derive(Debug, Display)]
+pub struct BackfillingStream<S: BlockStream + 'static, F: BlockFetcher + 'static> {
+    upstream: S,
+    fetcher: Arc<F>,
+    logger: Logger,
+}
+
+impl<S: BlockStream + 'static, F: BlockFetcher + 'static> BackfillingStream<S, F> {
+    /// Instantiate a [BackfillingStream].
+    pub fn new(upstream: S, fetcher: F, logger: Logger) -> Self {
+        Self {
+            upstream,
+            fetcher: Arc::new(fetcher),
```

Comment:
Perhaps, move the Arc creation to the stream object
```suggestion
            fetcher,
```